extensions/ql-vscode/src/queries-panel/query-discovery.ts

```diff
@@ -1,147 +1,245 @@
 import { dirname, basename, normalize, relative } from "path";
 import { Discovery } from "../common/discovery";
 import { CodeQLCliServer } from "../codeql-cli/cli";
-import { Event, RelativePattern, Uri, WorkspaceFolder } from "vscode";
+import {
+  Event,
+  RelativePattern,
+  Uri,
+  WorkspaceFoldersChangeEvent,
+  workspace,
+} from "vscode";
 import { MultiFileSystemWatcher } from "../common/vscode/multi-file-system-watcher";
 import { App } from "../common/app";
-import { FileTreeDirectory, FileTreeLeaf } from "../common/file-tree-nodes";
+import {
+  FileTreeDirectory,
+  FileTreeLeaf,
+  FileTreeNode,
+} from "../common/file-tree-nodes";
 import { getOnDiskWorkspaceFoldersObjects } from "../helpers";
 import { AppEventEmitter } from "../common/events";
 import { QueryDiscoverer } from "./query-tree-data-provider";
 import { extLogger } from "../common";
+import { FilePathQueue } from "./file-path-queue";
+import { QueryPack, QueryPackDiscovery } from "./query-pack-discovery";
+import { exists, lstat } from "fs-extra";
+import { isEqualOrHigherPath } from "../pure/helpers-pure";
 
-/**
- * The results of discovering queries.
- */
-export interface QueryDiscoveryResults {
-  /**
-   * A tree of directories and query files.
-   * May have multiple roots because of multiple workspaces.
-   */
-  queries: Array<FileTreeDirectory<string>>;
+const QUERY_EXTENSION = ".ql";
+const QUERY_REGEX = "**/*.ql";
 
-  /**
-   * File system paths to watch. If any ql file changes in these directories
-   * or any subdirectories, then this could signify a change in queries.
-   */
-  watchPaths: Uri[];
+interface Query {
+  path: string;
+  language: string | undefined;
 }
 
 /**
  * Discovers all query files contained in the QL packs in a given workspace folder.
  */
 export class QueryDiscovery
```

Comment:
It's occurred to me that there's a lot of duplication between these two discovery classes in the way they have the queue of paths and maintain an array of known files. Very likely a shared base class could be pulled out.